File: homework1/benjamin_silva/q4_BackspaceStringCompare.py
def backspaceCompare(s, t):
    S = len(s) - 1
    T = len(t) - 1
    skip_counterS = 0
    skip_counterT = 0
    while S >= 0 or T >= 0:
        while S >= 0 and s[S] == '#':
            skip_counterS += 1
            S -= 1
        while S >= 0 and skip_counterS > 0 and s[S] != '#':
            skip_counterS -= 1
            S -= 1
        while T >= 0 and t[T] == '#':
            skip_counterT += 1
            T -= 1
        while T >= 0 and skip_counterT > 0 and t[T] != '#':
            skip_counterT -= 1
            T -= 1
        if (S < 0 and T >= 0) or (S >= 0 and T < 0):
            return False
        if S < 0 and T < 0:
            return True
        if S >= 0 and T >= 0 and s[S] != t[T]:
            return False
        else:
            S -= 1
            T -= 1
    return True

def test_backspaceCompare():
    assert backspaceCompare("abcde", "abcde") == True, "Test 1 Failed"

    assert backspaceCompare("Uber Career Prep", "u#Uber Careee#r Prep") == True, "Test 2 Failed"

    assert backspaceCompare("abcdef###xyz", "abcw#xyz") == True, "Test 3 Failed"

    assert backspaceCompare("abcdef###xyz", "abcdefxyz###") == False, "Test 4 Failed"

if __name__ == "__main__":
    test_backspaceCompare()

File: homework1/rodrigo_canto/q4_BackSpaceStringCompare.cpp
// Technique: Stack

// Time Complexity: O(N)
// Space Complexity: O(N)

#include <bits/stdc++.h>
using namespace std;

//We can simulate the backspace effect using a stack.

stack<char> process(const string &s){

    stack<char> result;

    for(char character : s){
        
        //if we find a backspace we erase the last character written, otherwise we insert the character at the top of the stack.
        if(character == '#'){
            if(!result.empty()){
                result.pop();
            }
        }
        else{
            result.push(character);
        }
    }

    return result;

}

double solve(const string &a, const string &b){

    stack<char> first_string = process(a);
    stack<char> second_string = process(b);

    //If the content for both stacks is the same, the result is the same.

    if(first_string.size() != second_string.size()) return false;

    while(!first_string.empty()){

        char first_char = first_string.top();
        char second_char = second_string.top();

        if(first_char != second_char) return false;
        
        first_string.pop();
        second_string.pop();
    }

    return true;
}

int main() {

    vector<pair<string, string>> cases;

    pair<string, string> case1 = make_pair("abcde", "abcde");
    pair<string, string> case2 = make_pair("Uber Career Prep", "u#Uber Careee#r Prep");
    pair<string, string> case3 = make_pair("abcdef###xyz", "abcw#xyz");
    pair<string, string> case4 = make_pair("abcdef###xyz", "abcdefxyz###");


    cases.emplace_back(case1);
    cases.emplace_back(case2);
    cases.emplace_back(case3);
    cases.emplace_back(case4);


    for(auto &test_case: cases){
        
        string a = test_case.first;
        string b = test_case.second;

        if(solve(a, b)) cout << "True" << "\n";
        else cout << "False" << "\n";
    }

    return 0;
}

// Time: 10 minutes

